src/kinesis/producer.py

```diff
@@ -99,47 +119,133 @@ def loop(self):
         # the loop() method to ensure that we've drained any remaining items from our queue before we exit.
-        self.buffer_time = 0.25
+        self.buffer_time = 0.001
+        self.ending = True
         self.loop()
 
+        while self.records:
+            self.flush_records()
+
+    def send_records(self, records):
+        try:
+            response = self.client.put_records(
+                    StreamName=self.stream_name,
+                    Records=records
+            )
+            failed_record_count = response['FailedRecordCount']
+
+            if failed_record_count > 0:
+                log.error(dict(msg='There were %i failures in a batch of %i' % (failed_record_count, len(records)), type='kinesis_send_failure'))
+                response_records = response['Records']
+                failed_records = []
+                for index, record in enumerate(response_records):
+                    if 'ErrorCode' in record:
+                        failed_records.append(records[index])
+                return False, failed_records
+            else:
+                return True, None
+        except Exception:
+            log.exception(dict(msg='Could not send records to Kinesis', type='kinesis_send_error'))
+            return False, records
+
     def flush_records(self):
         if self.records:
-            log.debug("Flushing %d records", len(self.records))
-            self.client.put_records(
-                StreamName=self.stream_name,
-                Records=self.records
-            )
+            log.debug(dict(msg="Flushing %d records" % len(self.records), type='kinesis_batch_flushing'))
+            should_send_records = True
+            records_to_send = self.records
+            retries = 0
+            while should_send_records:
+                success, failed_records = self.send_records(records_to_send)
+
+                if success:
+                    should_send_records = False
+                else:
+                    if retries >= self.max_retry_count:
+                        log.error(dict(msg='Exceeded the maximum number of retries while trying to send records to Kinesis.  Dropping %i records' % len(failed_records), type='kinesis_max_retry_exceeded'))
+                        self.log_unsent_data(failed_records)
+                        should_send_records = False
+                    else:
+                        records_to_send = failed_records
+                        retries += 1
+                        # sleep a random amount so that we do not bombard Kinesis
+                        sleep_time = random.random() * 2.0
+                        log.debug(dict(msg='Failed sending %i records to Kinesis.  Sleeping for %f seconds' % (len(failed_records), sleep_time), type='kinesis_failure_sleep'))
+                        time.sleep(sleep_time)
+
+        self.set_next_batch()
+
+    def set_next_batch(self):
```

Comment:
It seems like this function is very similar in purpose to `get_records`. What is the reason for having both versions and why do we need to call this at the end of `flush_records` when `get_records` will be called at the beginning of the next loop?

Reply:
There's two buffers at play here -- the `records` and `next_records` instance variables.  When `get_records` is being called, we're draining records from the queue.  When we're doing this, we can have a case where we drain more records than should be sent at one time (as defined by `max_count` and `max_size`).  This generally happens when the record we just drained puts us over the `max_size` value or if we're ending operation and we want to drain all records that were put in the queue.  We try and make sure that the `records` instance variable is a sendable batch (again, defined by `max_count` and `max_size`) and then just add the newly drained record to the `next_records` buffer.  

We call `set_next_batch` at the end of `flush_records` to move records from `next_records` to `records`.  It is better to do this here, rather than within `get_records` as we may not be calling `get_records` in the case where we are ending operation.  In the `end` method, we keep calling `flush_records` while `records` is non-empty.  We don't want to call `get_records` here since we don't want to keep trying to read from the queue.

Reply:
Ok, so these changes are also modifying the ending behavior to only flush records that have been pulled off the queue, instead of trying to drain the whole queue like before. Can you update the comment in `end` to reflect that?

Reply:
So the comment actually better represents what is now happening vs. what was previously happening.  With my changes, we're draining the whole queue.  Previously, we would only drain until we reached the buffer timeout.  So if draining the queue took more than 250ms, we would stop there and then flush.  

The `end` method calls `loop`, which will call `get_records` to fully drain the queue.  If there are more records than would fit into one batch, then we put those records in `next_records`.  The `loop` method will then call `flush_records`, which will send what's currently in `records` and then call `set_next_batch` to move records from `next_records` to `records`.  In `end`, that's where the `while self.records` loop comes into play so that we can ensure that we send all the records that have been drained from the queue.

Reply:
Ok I think I get it now. It's a little confusing but it makes sense